specs/ChainSpec/SelectingBlockProducers.md

```diff
@@ -0,0 +1,335 @@
+
+# smallest value of s_min / S such that
+# (1 - (s_min / S))^epoch_length < PROBABILITY_NEVER_SELECTED
+min_stake_fraction =
+    1 - PROBABILITY_NEVER_SELECTED^(1/epoch_length)
+
+# we assume the stake from chunk producers will be roughly
+# evenly distributed among all the shards, so the stake fraction
+# we care about is really smaller than the quantity above.
+min_stake_fraction /= num_shards
+
+chunk_producers = []
+total_stake = 0
+for v in sorted_proposals[0:(MAX_NUM_BP + MAX_NUM_CP)]:
+    total_stake += v.stake
+    if (v.stake / total_stake) > min_stake_fraction:
+        chunk_producers.append(v)
+    else:
+        break
+
+chunk_producers.sort_descending(key=lambda v: (v.stake, v.account_id))
+# using the algorithm above to assign shards
+validator_shard_assignments =
+    assign_chunk_producers(chunk_producers, num_shards, min_validators_per_shard)
+
+chunk_producer_sampler = [
+    WeightedIndex([v.stake for v in shard_cps]) for shard_cps in validator_shard_assignments
+]
+
+return (chunk_producers, validator_shard_assignments, chunk_producer_sampler)
+```
+
+## Algorithm for sampling validators proportional to stake
+
+This algorithm is applied using both chunk-only producers and block producers in the subsequent
+algorithms for selecting a specific block producer and chunk producer at each height.
+
+### Input
+
+* `rng_seed: [u8; 32]`
+  - See usages of this algorithm below to see how this seed is generated
+* `validators: Vec<ValidatorStake>`
+* `sampler: WeightedIndex`
+
+### Output
+
+* `selection: ValidatorStake`
+
+### Steps
+
+```python
+# The seed is used as an entropy source for the random numbers.
+# The first 8 bytes select a block producer uniformly.
+uniform_index = int.from_bytes(rng_seed[0:8], byteorder='little') % len(validators)
+
+# The next 16 bytes uniformly pick some weight between 0 and the total
+# weight (i.e. stake) of all block producers.
+let uniform_weight = int.from_bytes(rng_seed[8:24], byteorder='little') \
+    % sampler.weight_sum()
+
+# Return either the uniformly selected block producer, or its "alias"
+# depending on the uniformly selected weight.
+index = uniform_index \
+    if uniform_weight < sampler.no_alias_odds[uniform_index] \
+    else sampler.aliases[uniform_index]
+
+return validators[index]
+```
+
+## Algorithm for selecting producer of block at height \\( h \\)
+
+### Input
+
+* `h: BlockHeight`
+  - Height to compute the block producer for
+  - Only heights within the epoch corresponding to the given block producers make sense as input
+* `block_producers: Vec<ValidatorStake>` (output from above)
+* `block_producer_sampler: WeightedIndex`
+* `epoch_rng_seed: [u8; 32]`
+  - Fixed seed for the epoch determined from Verified Random Function (VRF) output of last block in
+    the previous epoch
+
+### Output
+
+* `block_producer: ValidatorStake`
+
+### Steps
+
+```python
+# Concatenates the bytes of the epoch seed with the height,
+# then computes the sha256 hash.
```

Comment:
Why do we need a cryptographically secure hash here?

Reply:
I think it is needed so that it becomes computationally infeasible to construct an `epoch_seed` which gives rise to a pre-defined block/chunk producer order. Otherwise, an attacker may try to influence the VRF entropy in order to get a particular sequence, e.g. to have members of a byzantine cartel make many blocks in a row to cause some problems in the network.

Reply:
That is quite difficult in practice. Given that each block producer has one bit of influence on the vrf output, if an attacker wants to achieve something meaningful, they need to have a large amount of stake that would allow them to initiate other attacks more easily. In addition, you cannot predict when the epoch is going to until shortly before it ends and therefore, it is quite difficult for an attacker to influence the vrf output of the last block of an epoch.

Reply:
> each block producer has one bit of influence on the vrf output

It's one bit per block right? And since BPs will be chosen in proportion to their stake this means they have influence proportional to their stake as well. E.g. if a cartel had 1/8 (12.5%) of the total stake then they could influence ~32 bits in the next epoch seed. Obviously this would not be enough to control the entire sequence of BPs in the next epoch, but I assume it means it would be possible to control a non-trivial sub-sequence (say 10 blocks long), where that sub-sequence my occur at any point during the next epoch (i.e. I don't think you could control the first 10 blocks for example, that is too narrow, but controlling _some_ run of 10 blocks seems feasible). I haven't done any experiment to try this for myself, this is only my intuition, and so may not be at all correct.

> That is quite difficult in practice

I agree. Though, I view that argument as falling in the category of "security by obscurity", which I don't think is a very strong guarantee. If something is hard then there will need to be a large incentive for someone to attempt it. But on the blockchain, especially a successful one, there do tend to be large incentives, and it might be that someone thinks the payout for pulling off a difficult attack is worth the effort.

Is your main concern with using a cryptographic hash performance? If so, we do not need to use sha256. We could use a faster cryptographic hash function; I only chose sha256 because it is what we use everywhere else in the system.

Reply:
Actually using sha256 is fine. We can just cache the result